### PR TITLE
Implement datetime downcasting for accurate column comparison in constraints

### LINKED ISSUE #2275: Inequality constraint cannot be applied to compare datetime to date
### Environment Details
* SDV version: 1.17.1 (latest)

### Error Description
_This bug was first noticed by a Slack user._

In my table, I have a datetime column (`submission_timestamp`) and a date column (`due_date`). I want to synthesize data with an Inequality constraint showing that `submission_timestamp <= due_date`.

image

However, I am unable to apply an Inequality constraint to this data; SDV complains that the data violates the constraint.

The problem is that the date does not have enough granularity;
- SDV (and Python in general) assumes that a due date such as `2016-10-12` is referring to the beginning of the day (`2016-10-12 00:00:00`)
- However, in my dataset, a due date of `2016-10-12` is referring to the _end of the day_ (`2016-10-12 11:59:59`) because it is ok to make a submission any time exactly on that day. This is why the inequality `submission_timestamp <= due_date` should be true.

### Expected Behavior
I expect that when `strict_boundaries=False` with a date column, the assumed timestamp allow for the loosest possible interpretation:
- If a date is a high column, we should assume it is referring to the _end of day_ 
- If a date is a low column, we should assume it is referring to the _beginning of day_

The opposite should be assumed when `strict_boundaries=True`.

### Steps to reproduce

```python
import pandas as pd

from sdv.metadata import Metadata
from sdv.single_table import GaussianCopulaSynthesizer

data = pd.DataFrame(data={
    'SUBMISSION_TIMESTAMP': ['2016-07-10 17:04:00', '2016-07-11 13:23:00', '2016-07-12 08:45:30', '2016-07-11 12:00:00', '2016-07-12 10:30:00'],
    'DUE_DATE': ['2016-07-10', '2016-07-11', '2016-07-12', '2016-07-13', '2016-07-14'],
    
})

metadata = Metadata.load_from_dict({
    'tables': {
        'table': {
            'columns': {
                'SUBMISSION_TIMESTAMP': { 'sdtype': 'datetime', 'datetime_format': '%Y-%m-%d %H:%M:%S' },
                'DUE_DATE': { 'sdtype': 'datetime', 'datetime_format': '%Y-%m-%d'}
            }
        }
    }
})
synthesizer = GaussianCopulaSynthesizer(metadata)

constraint = {
    'constraint_class': 'Inequality',
    'constraint_parameters': {
        'low_column_name': 'SUBMISSION_TIMESTAMP',
        'high_column_name': 'DUE_DATE',
        'strict_boundaries': False
    }
}

synthesizer.add_constraints([constraint])

synthesizer.fit(data)
synthesizer.sample(num_rows=10)
```
```
ConstraintsNotMetError: 
Data is not valid for the 'Inequality' constraint:
  SUBMISSION_TIMESTAMP    DUE_DATE
0  2016-07-10 17:04:00  2016-07-10
1  2016-07-11 13:23:00  2016-07-11
2  2016-07-12 08:45:30  2016-07-12
```

### PR DESCRIPTION
**Pull request by pvk-developer**
Resolves #2275 
CU-86b2q1p22

### CODE REVIEW

**Comment by sdv-team** [comment]
Task linked: [CU-86b2q1p22 SDV - Inequality constraint cannot be applied to compare datetime to date #2275](https://app.clickup.com/t/86b2q1p22)

File: sdv/constraints/utils.py [hunk 1]
```diff
@@ -221,3 +257,98 @@ def get_mappable_combination(combination):
             A mappable combination of values.
     """
     return tuple(None if pd.isna(x) else x for x in combination)
+
+
+def match_datetime_precision(low, high, low_datetime_format, high_datetime_format):
+    """Match `low` or `high` datetime array to match the lower precision format.
```
Comment (amontanez24) [inline sdv/constraints/utils.py:263]:
```suggestion
    """Match `low` or `high` datetime array to the lower precision format.
```

File: sdv/constraints/utils.py [hunk 2]
```diff
@@ -221,3 +257,98 @@ def get_mappable_combination(combination):
...
+    """
+    primary_level = get_datetime_format_precision(primary_format)
+    secondary_level = get_datetime_format_precision(secondary_format)
+    if primary_level >= secondary_level:
+        return secondary_format
+
+    return primary_format
+
+
+def downcast_datetime_to_lower_format(data, target_format):
```
Comment (amontanez24) [inline sdv/constraints/utils.py:323]: minor: I'd call it `downcast_datetime_to_lower_precision`

File: sdv/constraints/utils.py [hunk 3]
```diff
@@ -221,3 +257,98 @@ def get_mappable_combination(combination):
...
+            The datetime format of the `high` column.
+
+    Returns:
+        Tuple[np.ndarray, np.ndarray]:
+            Adjusted `low` and `high` arrays where the higher precision format is
+            downcasted to the lower precision format.
+    """
+    lower_precision_format = get_lower_precision_format(low_datetime_format, high_datetime_format)
+    if lower_precision_format == high_datetime_format:
+        low = downcast_datetime_to_lower_format(low, lower_precision_format)
```
Comment (amontanez24) [inline sdv/constraints/utils.py:282]: codecov says this is untested?

File: sdv/constraints/tabular.py [hunk 4]
```diff
@@ -484,6 +485,14 @@ def is_valid(self, table_data):
             low = cast_to_datetime64(low, self._low_datetime_format)
             high = cast_to_datetime64(high, self._high_datetime_format)
 
+            if self._low_datetime_format != self._high_datetime_format:
+                low, high = match_datetime_precision(
```
Comment (amontanez24) [inline sdv/constraints/tabular.py:489]: Idk if this is a costly operation but I think we only needed to do this when `strict_boundaries` is False

**Review by amontanez24 — APPROVED** [review]
LGTM!

**Review by frances-h — APPROVED** [review]

**Pull request merged**